src/Plugin/GraphQL/DataProducer/Field/EntityReferenceRevisions.php

```diff
@@ -0,0 +1,214 @@
+   * @param bool $access
+   *   Whether check for access or not. Default is true.
+   * @param \Drupal\Core\Session\AccountInterface|null $accessUser
+   *   User entity to check access for. Default is null.
+   * @param string $accessOperation
+   *   Operation to check access for. Default is view.
+   * @param \Drupal\graphql\GraphQL\Execution\FieldContext $context
+   *   The caching context related to the current field.
+   *
+   * @return \GraphQL\Deferred|null
+   *   A promise that will return entities or NULL if there aren't any.
+   */
+  public function resolve(EntityInterface $entity, string $field, ?string $language, ?array $bundles, ?bool $access, ?AccountInterface $accessUser, ?string $accessOperation, FieldContext $context): ?Deferred {
+    if (!$entity instanceof FieldableEntityInterface || !$entity->hasField($field)) {
+      return NULL;
+    }
+
+    $definition = $entity->getFieldDefinition($field);
+    if ($definition->getType() !== 'entity_reference_revisions') {
+      return NULL;
+    }
+
+    $definition = $entity->getFieldDefinition($field);
+    $type = $definition->getSetting('target_type');
+    if (($values = $entity->get($field)) && $values instanceof EntityReferenceFieldItemListInterface) {
+      $vids = array_map(function ($value) {
+        return $value['target_revision_id'];
+      }, $values->getValue());
+
+      $resolver = $this->entityRevisionBuffer->add($type, $vids);
+      return new Deferred(function () use ($type, $language, $bundles, $access, $accessUser, $accessOperation, $resolver, $context) {
+        $entities = $resolver() ?: [];
+        $entities = array_filter($entities, function (EntityInterface $entity) use ($bundles, $access, $accessOperation, $accessUser, $context) {
+          if (isset($bundles) && !in_array($entity->bundle(), $bundles)) {
+            return FALSE;
+          }
+
+          // Check if the passed user (or current user if none is passed) has
+          // access to the entity, if not return NULL.
+          if ($access) {
+            /* @var $accessResult \Drupal\Core\Access\AccessResultInterface */
+            $accessResult = $entity->access($accessOperation, $accessUser, TRUE);
+            $context->addCacheableDependency($accessResult);
+            if (!$accessResult->isAllowed()) {
+              return FALSE;
+            }
+          }
+
+          return TRUE;
+        });
+
+        if (empty($entities)) {
+          $type = $this->entityTypeManager->getDefinition($type);
+          /** @var \Drupal\Core\Entity\EntityTypeInterface $type */
+          $tags = $type->getListCacheTags();
+          $context->addCacheTags($tags);
+          return NULL;
+        }
+
+        if (isset($language)) {
+          $entities = array_map(function (EntityInterface $entity) use ($language) {
+            if ($language != $entity->language()->getId() && $entity instanceof TranslatableInterface) {
+              return $entity->getTranslation($language);
```

Comment:
Access checks are usually per language, no? If true, then we need to actually invoke the access check after loading the translation, not before.

Reply:
Maybe worth checking the other data producers too if they do it correctly?